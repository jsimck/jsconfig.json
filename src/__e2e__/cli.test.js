const path = require('path');
const { spawn } = require('child_process');

async function runTest(name, cliArgs = [], cwd = null) {
  return new Promise((resolve, reject) => {
    const cli = spawn(
      'node',
      [
        path.resolve(__dirname, '../cli.js'),
        `--output=${path.resolve(__dirname, `./results/${name}`)}`,
        ...cliArgs,
        ...(cwd ? [cwd] : [])
      ],
      {
        // Set to 'inherit' for debugging
        stdio: 'inherit'
      }
    );

    cli.on('exit', () => {
      resolve(require(`./results/${name}/jsconfig.json`));
    });

    cli.on('error', (err) => {
      reject(err);
    });
  });
}

describe('jsconfig.json CLI', () => {
  it('should generate basic jsconfig.json', async () => {
    expect(await runTest('default')).toMatchSnapshot();
  });

  it('should work with custom cwd', async () => {
    expect(
      await runTest('aliases', [], './src/__e2e__/mocks/aliases')
    ).toMatchSnapshot();
  });

  it('should work with custom webpack location and baseUrl', async () => {
    expect(
      await runTest('customConfig', [
        '--baseUrl=./myApp/custom/location',
        `--webpackConfig=${path.resolve(
          __dirname,
          './mocks/customConfig/custom.webpack.js'
        )}`
      ])
    ).toMatchSnapshot();
  });

  it.each(
    ['default', 'nextjs', 'react', 'vuejs', 'node'].map((name) => [name])
  )('should generate jsconfig.json for %s template correctly', async (name) => {
    expect(await runTest(name, [`--template=${name}`])).toMatchSnapshot();
  });

  it('should merge path aliases with default template', async () => {
    expect(
      await runTest('templateOverride', [
        '--baseUrl=./myApp/custom/location',
        '--template=nextjs',
        `--webpackConfig=${path.resolve(
          __dirname,
          './mocks/customConfig/custom.webpack.js'
        )}`
      ])
    ).toMatchSnapshot();
  });

  it('should allow cli args overrides', async () => {
    expect(
      await runTest('cliArgsOverride', [
        '--baseUrl=./myApp/custom/location',
        '--template=nextjs',
        '--module=amd',
        '--target=es3',
        '--moduleResolution=classic',
        '--experimentalDecorators=false',
        '--syntheticImports=false',
        `--webpackConfig=${path.resolve(
          __dirname,
          './mocks/customConfig/custom.webpack.js'
        )}`
      ])
    ).toMatchSnapshot();
  });
});
