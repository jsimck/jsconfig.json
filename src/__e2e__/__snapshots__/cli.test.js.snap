// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsconfig.json CLI should allow cli args overrides 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": "./myApp/custom/location",
    "checkJs": false,
    "module": "amd",
    "moduleResolution": "classic",
    "paths": Object {
      "@/components/*": Array [
        "components/*",
      ],
      "@/pages/*": Array [
        "pages/*",
      ],
      "@/styles/*": Array [
        "styles/*",
      ],
      "myApp/*": Array [
        "../../../src/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es3",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".next",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
    "include": Array [
      "react",
      "react-dom",
    ],
  },
}
`;

exports[`jsconfig.json CLI should generate basic jsconfig.json 1`] = `
Object {
  "compilerOptions": Object {
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".nuxt",
    "coverage",
    "jspm_packages",
    "tmp",
    "temp",
    "bower_components",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
  },
}
`;

exports[`jsconfig.json CLI should generate jsconfig.json for default template correctly 1`] = `
Object {
  "compilerOptions": Object {
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".nuxt",
    "coverage",
    "jspm_packages",
    "tmp",
    "temp",
    "bower_components",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
  },
}
`;

exports[`jsconfig.json CLI should generate jsconfig.json for nextjs template correctly 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "paths": Object {
      "@/components/*": Array [
        "components/*",
      ],
      "@/pages/*": Array [
        "pages/*",
      ],
      "@/styles/*": Array [
        "styles/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".next",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
    "include": Array [
      "react",
      "react-dom",
    ],
  },
}
`;

exports[`jsconfig.json CLI should generate jsconfig.json for node template correctly 1`] = `
Object {
  "compilerOptions": Object {
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
    "include": Array [
      "node",
    ],
  },
}
`;

exports[`jsconfig.json CLI should generate jsconfig.json for react template correctly 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": "src",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
    "include": Array [
      "react",
      "react-dom",
      "jest",
      "testing-library__jest-dom",
    ],
  },
}
`;

exports[`jsconfig.json CLI should generate jsconfig.json for vuejs template correctly 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "paths": Object {
      "@/*": Array [
        "./src/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".nuxt",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
  },
}
`;

exports[`jsconfig.json CLI should merge path aliases with default template 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": "./myApp/custom/location",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "paths": Object {
      "@/components/*": Array [
        "components/*",
      ],
      "@/pages/*": Array [
        "pages/*",
      ],
      "@/styles/*": Array [
        "styles/*",
      ],
      "myApp/*": Array [
        "../../../src/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".next",
    "coverage",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
    "include": Array [
      "react",
      "react-dom",
    ],
  },
}
`;

exports[`jsconfig.json CLI should work with custom cwd 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "paths": Object {
      "myApp/*": Array [
        "src/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".nuxt",
    "coverage",
    "jspm_packages",
    "tmp",
    "temp",
    "bower_components",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
  },
}
`;

exports[`jsconfig.json CLI should work with custom webpack location and baseUrl 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": "./myApp/custom/location",
    "checkJs": false,
    "module": "es2015",
    "moduleResolution": "node",
    "paths": Object {
      "myApp/*": Array [
        "../../../src/*",
      ],
    },
    "resolveJsonModule": true,
    "target": "es2020",
  },
  "exclude": Array [
    "dist",
    "node_modules",
    "build",
    ".vscode",
    ".nuxt",
    "coverage",
    "jspm_packages",
    "tmp",
    "temp",
    "bower_components",
    ".npm",
    ".yarn",
  ],
  "typeAcquisition": Object {
    "enable": true,
  },
}
`;
